src/puya/awst/nodes.py

```diff
@@ -962,6 +962,29 @@ def accept(self, visitor: ExpressionVisitor[T]) -> T:
         return visitor.visit_tuple_item_expression(self)
 
 
+@attrs.frozen
+class NamedTupleExpression(Expression):
+    """
+    Defines a new named tuple with the specified elements.
+    Elements are evaluated in mapping order.
+    """
+
+    wtype: wtypes.WTuple
+    values: Mapping[str, Expression] = attrs.field(converter=immutabledict)
+
+    @values.validator
+    def _validate_values(self, _instance: object, values: Mapping[str, Expression]) -> None:
+        if values.keys() != self.wtype.fields.keys():
+            raise CodeError("Invalid argument(s)", self.source_location)
```

Comment:
I can see this uses the same code as NewStruct which is good, however we're in the process of making error messages start with lower case (this is rather inconsistent at the moment, but there are some larger changes planned here around error reporting for the language server, and warning treatments as well generally, so this work is more of a fix-it-as-you-go thing).